File: luogu/Monthly/2/t1.cpp
#include <cstdio>
#include <cstring>
#include <cstdlib>
using namespace std;

const int maxn=20;
int seedA[3], seedB[3];
int count[maxn][maxn];
long double dis[maxn][maxn];

void makeData(int n, int seedA[], int seedB[]) {
	long long x=0, y=0, z=0;
	double a, b; 
	const int rp = 2333333;
	for (int i=0; i<n; i++) {
		x = (y * seedA[0] + seedB[0]) % rp;
		y = (z * seedA[1] + seedB[1]) % rp;
		z = (x * seedA[2] + seedB[2]) % rp;
		a = x % 20 + y % 10 / 10.0 + z % 10 / 100.0;

		x = (y * seedA[0] + seedB[0]) % rp;
		y = (z * seedA[1] + seedB[1]) % rp;
		z = (x * seedA[2] + seedB[2]) % rp;
		b = x % 20 + y % 10 / 10.0 + z % 10 / 100.0;

		dis[int(a)][int(b)] += a + b - int(a) - int(b);
		++count[int(a)][int(b)];
	}
}

int main() {
	int n;
	scanf("%d", &n);
	for (int i=0; i<3; i++) scanf("%d %d", seedA+i, seedB+i);
	memset(dis, 0, sizeof(dis));
	memset(count, 0, sizeof(count));
	makeData(n, seedA, seedB);
	double sum = 0.0, num = 0.0;
	for (int x1=0; x1<maxn; x1++) for (int y1=0; y1<maxn; y1++)
	for (int x2=x1+1; x2<maxn; x2++) for (int y2=y1+1; y2<maxn; y2++) {
		num += count[x1][y1] * count[x2][y2];
		sum += (x2+y2-x1-y1) * count[x1][y1] * count[x2][y2] \
		     - dis[x1][y1] * count[x2][y2] + dis[x2][y2] * count[x1][y1];
	}
	if (num == 0.0) printf("0.00000\n");
	else printf("%.5lf\n", sum/num);
	return 0;
}

File: luogu/Monthly/2/t1.pas
program t1;
const
  maxn=20;
var
  dis:array [0..maxn, 0..maxn] of double;
  count:array [0..maxn, 0..maxn] of longint;
  x1, x2, y1, y2:longint;
  sum, num:double;

procedure makeData();
  const rp=2333333;
  var
    n, i:longint;
    x, y, z:int64;
    a, b:double;
    sa, sb:array[1..3] of longint;
  begin
    readln(n);
    for i:=1 to 3 do read(sa[i], sb[i]);
    readln;
    x:=0; y:=0; z:=0;
    for i:=1 to n do
    begin
      x:=(y * sa[1] + sb[1]) mod rp;
      y:=(z * sa[2] + sb[2]) mod rp;
      z:=(x * sa[3] + sb[3]) mod rp;
      a:=x mod 20 + y mod 10 / 10 + z mod 10 / 100;

      x:=(y * sa[1] + sb[1]) mod rp;
      y:=(z * sa[2] + sb[2]) mod rp;
      z:=(x * sa[3] + sb[3]) mod rp;
      b:=x mod 20 + y mod 10 / 10 + z mod 10 / 100;

      dis[trunc(a), trunc(b)]:=dis[trunc(a), trunc(b)]+ a+b-trunc(a)-trunc(b);
      inc(count[trunc(a), trunc(b)]);
    end;
  end;

begin
  makeData;
  sum:=0.0; num:=0.0;
  for x1:=0 to maxn-2 do for y1:=0 to maxn-2 do
  for x2:=x1+1 to maxn-1 do for y2:=y1+1 to maxn-1 do
  begin
    num:=num+count[x1, y1]*count[x2, y2];
    sum:=sum + (x2+y2-x1-y1)*count[x1, y1]*count[x2, y2] - dis[x1, y1]*count[x2, y2] + dis[x2, y2]*count[x1, y1];
  end;
  if num=0 then writeln(0.0:0:5)
  else writeln(sum/num:0:5);
end.